// genex/build/iterators/Stack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Literal_1 = require("./Literal");
class Stack {
    constructor(data) {
        if (data.length === 0) {
            data = [new Literal_1.Literal([''])];
        }
        this.data = data;
    }
    *[Symbol.iterator]() {
        for (this.rewind(); this.valid() === true; this.next()) {
            yield this.current();
        }
    }
    clone() {
        return new Stack(this.data.map((iterator) => iterator.clone()));
    }
    current() {
        let result = '';
        for (let value of this.data) {
            result += value.current();
        }
        return result;
    }
    next() {
        if (this.valid() === true) {
            let i = this.data.length;
            while ((this.data[--i].next(), i > 0 && this.data[i].valid() !== true)) {
                this.data[i].rewind();
            }
        }
    }
    rewind() {
        for (let value of this.data) {
            value.rewind();
        }
    }
    valid() {
        return this.data.length > 0 && this.data[0].valid();
    }
}
exports.Stack = Stack;
